### PR TITLE
[CA-58] Add deploy script

### PR DESCRIPTION
**Pull request by w3cj**
* Updates .travis.yml to deploy after success

### CODE REVIEW

**Comment by w3cj** [comment]
DO NOT MERGE. Need to work out the deploy.sh script

**Comment by snordmann** [comment]
I have several questions about the choices for this PR.

Why don't we use the build system of TravisCI? 
We can use conditionals to be sure to only deploy on branches that match certain names or regular expressions. Environment variables can be set there and we would not need to build a resilient deploy script.

Why do we use `/bin/sh` in the deploy script?
Using shell - in contrast to `/bin/bash` - is highly restrictive. You have used the double bracket regex syntax, that is only available in bash. Does TravisCI not support bash?

**Comment by w3cj** [comment]
@sbibow Yeah the deploy script is failing right now. Didn't realize sh doesn't support double bracket regex.

I'm fine with using the build system of travis, I'm just not familiar enough with it and couldn't figure out how to do what I wanted to do from the docs. I'm open to suggestions!

**Comment by snordmann** [comment]
Okay, that was a handful of commits 😅 It should now all be set for review. @w3cj Can you check if everything makes sense and is okay with you?

**Comment by spiray** [comment]
@sbibow It seems that api-dev.coding.garden is giving that 502. See the devops channel on discord where there was a discussion on why that may be happening.

**Comment by snordmann** [comment]
@spiray Well, this branch is not yet merged to develop and thus not deployed to api-dev.coding.garden. However in [the last build that also had a deployment phase](https://travis-ci.org/CodingGardenCommunity/app-backend/builds/507503810) you can see that deployment per se worked. The job failed because I wanted to alias it to a sub domain that is not set up. You can still [check the logs](https://travis-ci.org/CodingGardenCommunity/app-backend/jobs/507503812) and [verify the final result](https://app-backend-e2waiuzuj.now.sh).

**Comment by w3cj** [comment]
Why do we need MONGO_SECRET and MONGO_URI?

I'm not sure how I feel about doing it with Travis vs the script I made. The script was a lot less code and less environment variables stored on travis.

**Comment by snordmann** [comment]
tbqh I have no idea how MongoDB works and how to connect to it. I assumed that you would need an address and then a secret for authentication and authorization. That's why I set these environment variables.
And instead of the environment variables stored on travis we would have used now.sh secrets. I don't know exactly how zeit.co handles secrets but I assume that they are set as environment variables in the executor. Then I don't see why we should manage secrets in two different places. That's why I moved everything to travis.

There are two downsides I see when using my approach:
1. Environment variables don't scale well
Once we need more environment variables the deploy script will be longer and longer.
2. Everyone needs to install now
I have simply added now as development dependency. This enables the deployment with travis and makes this much easier at the cost of having everyone to install now. This is 50MB and a new dependency, that *nobody* will use, because we have automated deployment.

Your approach was not too bad, but I like it more if we have different job stages ("test", "release", "deploy", etc.) for more granular control and not having to worry too much about script stability and resilience.

---

We could still use a deploy script with arguments for the environment and still keep the general Travis CI job structure. Then we wouldn't have downside 1 and maybe we can even install now during the script to also negate downside 2.

**Comment by w3cj** [comment]
I think we should take the second approach. Have a deploy job in the travis.yml file that matches all branches that should be deployed and then defers to a shell script we can maintain. The shell script can install now before deploying.

**Comment by snordmann** [comment]
I have now written a deploy script. It can be used both in a CI or manually if that somehow becomes a requirement (though it leaves artefacts when running manually).

---

Another related question: Should we deploy *every* branch for contributors to directly see the results? 
We could deploy it, not alias it and connect it to the development MongoDB. 
When doing this we could test new features in real time as it would actually run later. However the development database will be **very** fragile and might be deleted by whomever. This would also increase your costs on now.sh significantly. What do you think?

**Comment by w3cj** [comment]
I don't think we should deploy every pull request for testing. Just develop / staging / master. 

I'm about to push a commit that moves all the environment variables into the script. I'm open to suggestions / fixes.

I think we discussed this before about release branches, but I'm thinking instead of deploying vX.X.X branches, we deploy the staging branch after a release branch has been merged into it. Let me know if there is anything wrong with this. (That is how I coded the script.)

**Comment by snordmann** [comment]
Script seems to be alright, though I haven't tested it. The only thing that might not work is the assignment of the `MONGO_URI` variable. Just to be sure I would have put it in quotation marks, in the case bash somehow treats the `@` differently.

I don't personally like the staging branch. It is just a branch more that doesn't even do anything. I like it more to combine staging and release branch. That way we have a PR less to merge when releasing a new version.

File: deploy.sh [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    if [ -z "$DEPLOY_ALIAS" ]; then
+      DEPLOY_ALIAS=api-dev.coding.garden
+    fi
+    if [ -z "$MONGO_URI" ]; then
+      MONGO_URI=@community-app-db-dev
+    fi
+    ;;
+  staging)
+    if [ -z "$NODE_ENV" ]; then
+      NODE_ENV=development
```
Comment (snordmann) [inline deploy.sh:106]: Shouldn't this be production in the staging area?

Comment (w3cj) [reply deploy.sh:106]: I have it as `development` so we can debug the stack traces in the logs in staging if needed.

**Pull request merged**